### PR TITLE
fix(expr): correctly handle unicode for substr

### LINKED ISSUE #9065: bug(expr): `SUBSTR` of unicode produces error: "byte index _ is not a char boundary"
We should not calculate by bytes, but by unicode character:

To reproduce:
Risingwave:
```sql
=> select substr('Mér', 1, 2);
SSL SYSCALL error: EOF detected
```

PSQL:
```sql
=> select substr('Mér', 1, 2);
 substr 
--------
 Mé
(1 row)
```

Similar for `''Mér'::char(3)`

---

resources:

src/expr/src/vector_op/substr.rs:48:23
https://github.com/risingwavelabs/risingwave/blob/95ab15c4dfa3b4b15086b4692812589b1b50bfc8/src/expr/src/vector_op/substr.rs#L37
https://stackoverflow.com/questions/4249745/does-postgresql-varchar-count-using-unicode-character-length-or-ascii-character

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Close #9065.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Review by wangrunji0408 — APPROVED** [review]

File: src/expr/src/vector_op/substr.rs [hunk 1]
```diff
@@ -12,40 +12,52 @@
...
+    for char in substr {
+        writer.write_char(char).unwrap();
+    }
+
     Ok(())
 }
 
-// #[function("substr(varchar, 0, int32) -> varchar")]
+// #[function("substr(varchar, 1, int32) -> varchar")]
+// TODO: when is this function called?
```
Comment (xiangjinwu) [inline src/expr/src/vector_op/substr.rs:35]: This is unused and can be removed.

File: src/expr/src/vector_op/substr.rs [hunk 2]
```diff
@@ -12,40 +12,52 @@
...
-    // 0 <= begin <= s.len()
-    let begin = min(max(start, 0) as usize, s.len());
-    let end = (start.saturating_add(count).max(0) as usize).min(s.len());
-    writer.write_str(&s[begin..end]).unwrap();
+
+    let skip = start.saturating_sub(1).max(0) as usize;
+    let take = if start >= 1 {
+        count as usize
+    } else {
+        count.saturating_add(start.saturating_sub(1)) as usize
```
Comment (xiangjinwu) [inline src/expr/src/vector_op/substr.rs:53]: `count.saturating_add(start.saturating_sub(1))` could be negative, and `as usize` would reinterpret it as a large positive value.

Comment (BugenZhao) [reply src/expr/src/vector_op/substr.rs]: Added a `max(0)` here and some test cases.

**Review by jon-chuang — APPROVED** [review]
LGTM

**Review by xiangjinwu — APPROVED** [review]

**Pull request merged**